### PR TITLE
Manually implement `std::future::Future::poll` to improve performance

### PR DESCRIPTION
**Pull request by chuigda**
According to benchmarking, this has raised performance by 20%~30%. Still, we need to double check the safety/soundness of this solution.

### CODE REVIEW

File: pr47-core/src/vm/al31f/executor/async_executor.rs [hunk 1]
```diff
@@ -148,45 +152,101 @@ unsafe fn checked_exception_unwind_stack<A: Alloc>(
...
-        let end_time: std::time::Instant = std::time::Instant::now();
-        eprintln!("Time consumed: {}ms", (end_time - start_time).as_millis());
-    });
+pub struct VMThreadRunFunctionFut<'a, A: Alloc, const S: bool> {
+    thread: &'a mut VMThread<A>,
+    slice: StackSlice,
+    insc_ptr: usize,
 
-    get_vm!(thread).alloc.set_gc_allowed(true);
+    awaiting_promise: Option<Pin<Box<dyn Future<Output = AsyncReturnType>>>>
```
Comment (chuigda) [inline pr47-core/src/vm/al31f/executor/async_executor.rs]: Here comes the question: could we avoid using such a `Pin<Box>`?

Comment (CousinZe) [reply pr47-core/src/vm/al31f/executor/async_executor.rs]: I guess that futures about `Mutex`es should be mostly `Unpin`. The only problem is how could we accomplish type erasure.

Comment (chuigda) [reply pr47-core/src/vm/al31f/executor/async_executor.rs]: There's now one thing for sure: we could neither explicitly write out the type, nor could we erase it without allocation.

**Review by CousinZe — APPROVED** [review]
LGTM

**Pull request merged**